core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerState.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.circuitbreaker;
+
+import java.time.Duration;
+
+import com.linecorp.armeria.common.util.Ticker;
+
+/**
+ * The state of the circuit breaker.
+ */
+final class CircuitBreakerState {
+    private final CircuitState circuitState;
+    private final EventCounter counter;
+    private final Ticker ticker;
+    private final long timedOutTimeNanos;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param circuitState The circuit state
+     * @param timeoutDuration The max duration of the state
+     * @param ticker ticker to measure duration
+     * @param counter The event counter to use during the state
+     */
+    CircuitBreakerState(CircuitState circuitState, Ticker ticker,
+                        Duration timeoutDuration, EventCounter counter) {
+        this.circuitState = circuitState;
+        this.counter = counter;
+        this.ticker = ticker;
+
+        if (timeoutDuration.isZero() || timeoutDuration.isNegative()) {
+            timedOutTimeNanos = 0L;
+        } else {
+            timedOutTimeNanos = ticker.read() + timeoutDuration.toNanos();
+        }
+    }
+
+    public EventCounter counter() {
+        return counter;
+    }
+
+    /**
+     * Returns {@code true} if this state has timed out.
+     */
+    public boolean checkTimeout() {
+        return 0 < timedOutTimeNanos && timedOutTimeNanos <= ticker.read();
+    }
+
+    public boolean isOpen() {
+        return circuitState == CircuitState.OPEN;
+    }
+
+    public boolean isHalfOpen() {
+        return circuitState == CircuitState.HALF_OPEN;
+    }
+
+    public boolean isClosed() {
+        return circuitState == CircuitState.CLOSED;
+    }
```

Comment:
How about?
```java
public interface CircuitBreakerState {
    CircuitState circuitState();

    boolean isTimedOut();

    boolean canRequest();
}
```
The following code could be replace with `currentState.canRequest()`.
https://github.com/line/armeria/blob/5b384fbe27e7e6f9225d6db91cbb684d09dfbb5e/core/src/main/java/com/linecorp/armeria/client/circuitbreaker/NonBlockingCircuitBreaker.java#L129-L130

Reply:
Maybe this suggestion is related with this comment (https://github.com/line/armeria/pull/4012/files#r784728692) 

Before jumping to it, I want to make a consensus with your idea.

1) If we want to replace above lines to new `CircuitBreakerState`, `CircuitBreakerState` should have role for creating new `CircuitBreakerState`. It means it should have some parameters - timeout value for each states & event counter for each counters.

2) I wonder how far we should abstract of `CircuitBreakerState`. Because I feel most of logics related with current state (onSuccess, onFailrue) should be moved to `CircuitBreakerState` as well if we move above lines to `CircuitBreakerState`. Then it makes the role between `CircuitBreakerState` and `NonBlockingCircuitBreaker` ambiguous. So I want to make the role of the class clear from your suggestion.

Reply:
By the way, I have been thinking of the semantics of `canRequest()`. I agreed that the original name caused some misunderstandings.
The new `canRequest()` only returns the ability of sending a request that can not make sense if it does not perform atomically.
Let's say, `CircuitState` is `HALF_OPEN` which allows only one trial request a time.
If `checkTimeout()` returns true, all concurrent requests will think they can be sent at the moment.

So how about addressing Trustin's idea? https://github.com/line/armeria/pull/4012#discussion_r781974556
But you might not want to use `CircuitState` for your code. 
The `CircuitState` can not be the permission of a request but just a snapshot state of a `CircuitBreaker`. 
Hence, I recommend to refactor your code to hold the first trial result.
```java
boolean canRequest = cb.tryRequest();

if (!canRequest) {
   // codes incurs GC 
}

// other codes 

if (!canRequest) {
    throw new RuntimeException();
}

boolean success = false;
try {
    success = client.sendRequest();
} finally {
    if (success) {
        cb.onSuccess();
    } else {
        cb.onFailure();
    }
}
```

Reply:
Sure. then let me revert to the code 

1. just renaming `canRequest` to`tryRequest` 
2. Add API for checking `CircuitBreakerState` 
- Though it is not clear for the usage, but I think it can be useful to provide hint of current circuitBreakerState (though it is not atomic). (Just in case of logging) 

And we can find similar API from resilience4j 
- https://github.com/resilience4j/resilience4j/blob/master/resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreaker.java#L597

Reply:
> Add API for checking CircuitBreakerState

I thought it was `CircuitState`.